pingrid/impl.py

```diff
@@ -90,6 +79,89 @@ def error_fig(error_msg="error"):
 FuncInterp2d = Callable[[Iterable[np.ndarray]], np.ndarray]
 
 
+class ColorScale:
+    
+    def __init__(self, name, colors, scale=None):
+        self.name = name
+        self.colors = colors
+        if scale is None:
+            self.scale = list(np.arange(len(colors)))
+        else:
+           if (np.diff(scale) < 0).any():
+              raise Exception("scale must be monotically increasing")
+           elif len(colors) == len(scale):
+               self.scale = scale
+           else:
+               raise Exception("scale must be same length as colors")
+
+    def reversed(self, name=None):
+        if name is None:
+            name = self.name + "_r"
+        return ColorScale(name, self.colors[::-1], self.scale)
+
+    def rescaled(self, new_min, new_max):
+        cs_val = np.array(self.scale)
+        scale = (cs_val - cs_val[0]) * (new_max - new_min) / (cs_val[-1] - cs_val[0]) + new_min
+        return ColorScale(self.name, self.colors, scale)
+
+    def to_rgba_array(self, lutsize=256):
+        cs = self.rescaled(0, lutsize-1)
+        nc =  len(cs.scale)
+        colors = np.array(cs.colors)
+        # append output is not used but saves writing a condition dedicated to last color
+        delta_colors  = np.diff(colors, axis=0, append=np.expand_dims(colors[-1,:], 0))
+        delta_scale = np.diff(cs.scale, append=cs.scale[-1])
+        # Construct lutsize x 4 RGBA array
+        rgbaa = np.transpose(np.array([np.piecewise(
+            # Rescale lut indices to colors piece by piece
+            np.arange(lutsize),
+            # Rescale differently from one anchor point to the next
+            [np.arange(lutsize) >= cs.scale[i] for i in range(nc)],
+            # Rescaling is linear from one anchor to the next
+            [np.polynomial.polynomial.Polynomial(
+                  # Unless it's a discontinuity then there is no rescaling
+                  [colors[i, band], 0] if np.diff(cs.scale, append=cs.scale[-1])[i] == 0
+                      # Intercept and slope of the linear relation
+                      else [
+                          colors[i, band] - cs.scale[i] * delta_colors[i, band] / delta_scale[i],
+                          delta_colors[i, band] / delta_scale[i],
+                      ]
+             ) for i in range(nc)]
+            # This is the lambda version of the Polynomial
+            # just can't get it to work... Leaving it here for someone smarter than me
+            ##[lambda x: (colors[i, band]
+            ##    + 0 if np.diff(cs.scale, append=cs.scale[-1])[i] == 0 else ((x - cs.scale[i]) * delta_colors[i, band] / delta_scale[i])
+            ##) for i in range(nc)]
+        # Same rescaling to all color bands
+        ) for band in range(4)])).astype(int)
```

Comment:
Do we really need to iterate over the bands explicitly? Is there a way to write it as a single calculation over a 2D array (variable value and band) instead of four 1D arrays?

If not, it's fine. Just looking to be more concise. Explicit loops in numpy code can be a problem for performance, but in this particular case it's not going to matter because the arrays are small.

Reply:
so its colors that has bands and colors is used to define the Polynomial functions. But Polynomials takes only 1-d array so I'd need to either make a loop at this level, or to decompose x and reform it. Either way, I would need to use the lambda x notation and as soon as I use that even without making any changes, it's not working anymore... (it's going through the last function).